### PR TITLE
docs: add explanation of encrypted training and federated learning

### PR DESCRIPTION
**Pull request by andrei-stoian-zama**
Doc updates for CML 1.4.0:
- mention training on encrypted data 
- mention FL 
- add a page on encrypted training
- improve the Logistic Reg encrypted training notebook 

Closes https://github.com/zama-ai/concrete-ml-internal/issues/4049
Closes https://github.com/zama-ai/concrete-ml-internal/issues/4154
Closes https://github.com/zama-ai/concrete-ml-internal/issues/4153
Closes https://github.com/zama-ai/concrete-ml-internal/issues/4036

### CODE REVIEW

File: docs/built-in-models/training.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
+# Training on Encrypted Data
+
+Concrete ML offers the possibility to train [SGD Logistic Regression](../developer-guide/api/concrete.ml.sklearn.linear_model.md#class-sgdclassifier) on encrypted data. The [logistic regression training](../advanced_examples/LogisticRegressionTraining.ipynb) example shows this feature in action.
+
+This example shows how to instantiate a logistic regression model that trains on encrypted data:
+
+```python
+parameters_range = (-1.0, 1.0)
+
+sgd_clf_binary_simulate = SGDClassifier(
```
Comment (RomanBredehoft) [inline docs/built-in-models/training.md:10]: not sure why this model is named `sgd_clf_binary_simulate` . I'm mostly taking about this "simulate" part, but I think you can just name it `model` or `model_encrypted_fit` for simplicity

File: docs/built-in-models/training.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
+# Training on Encrypted Data
+
+Concrete ML offers the possibility to train [SGD Logistic Regression](../developer-guide/api/concrete.ml.sklearn.linear_model.md#class-sgdclassifier) on encrypted data. The [logistic regression training](../advanced_examples/LogisticRegressionTraining.ipynb) example shows this feature in action.
+
+This example shows how to instantiate a logistic regression model that trains on encrypted data:
```
Comment (RomanBredehoft) [inline docs/built-in-models/training.md:5]: `instantiate a logistic regression ` -> `instantiate an SGD classifier`

File: docs/built-in-models/training.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+To activate encrypted training simply set `fit_encrypted=True` in the constructor. If this value is not set, training is performed
+on clear data using `scikit-learn` gradient descent.
+
+Next, to perform the training on encrypted data, call the `fit` function with the `fhe="execute"` argument:
+
+<!--pytest-codeblocks:skip-->
+
+```python
+sgd_clf_binary_fhe.fit(X_binary, y_binary, fhe="execute")
```
Comment (RomanBredehoft) [inline docs/built-in-models/training.md:26]: I would keep the same name as defined above

File: docs/built-in-models/training.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+The `max_iter` parameter controls the number of batches that are processed by the training algorithm. Good values for this parameter are 8-64.
+
+The `parameters_range` parameter determines the initialization of the coefficients and the bias of the logistic regression. It is recommended to give values that are close to the min/max of the training data. It is also possible to normalize the training data so that it lies in the range $$[-1, 1]$$.
+
+## Capabilities and Limitations
+
+The logistic model that can be trained uses Stochastic Gradient Descent (SGD) and quantizes for data, weights, gradients and the error measure. It currently supports training 6-bit models, training both the coefficients and the bias.
+
+The `SGDClassifier` does not currently support training models with other values for the bit-widths. Second, the time to train the model
```
Comment (RomanBredehoft) [inline docs/built-in-models/training.md:45]: not sure why the second sentence starts with `Second`

**Review by RomanBredehoft — CHANGES_REQUESTED** [review]
thanks ! I have very few comments

also, if you want the associated issue to be closes you need to write `closes https://github.com/zama-ai/concrete-ml-internal/issues/4049` (complete link)

File: docs/built-in-models/training.md (hunk 1 above)
Comment (jfrery) [reply docs/built-in-models/training.md:10]: Agreed let's make codeblock very user friendly.

File: docs/built-in-models/training.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+{% hint style="info" %}
+Training on encrypted data provides the highest level of privacy but is slower than training on clear data. Federated learning is an alternative approach, where data privacy can be ensured through _differential privacy_ instead of encryption. Concrete ML
+can import linear models, including logistic regression, that are trained using federated learning using the [`from_sklearn` function](linear.md#pre-trained-models).
+
+{% endhint %}
+
+## Training configuration
+
+The `max_iter` parameter controls the number of batches that are processed by the training algorithm. Good values for this parameter are 8-64.
```
Comment (jfrery) [inline docs/built-in-models/training.md:37]: > 8-64

what is this ? I would just say "Default value is..."  I don't think we can have a rule here.

File: docs/built-in-models/training.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+The `parameters_range` parameter determines the initialization of the coefficients and the bias of the logistic regression. It is recommended to give values that are close to the min/max of the training data. It is also possible to normalize the training data so that it lies in the range $$[-1, 1]$$.
+
+## Capabilities and Limitations
+
+The logistic model that can be trained uses Stochastic Gradient Descent (SGD) and quantizes for data, weights, gradients and the error measure. It currently supports training 6-bit models, training both the coefficients and the bias.
+
+The `SGDClassifier` does not currently support training models with other values for the bit-widths. Second, the time to train the model
+is proportional to the number of features and the number of training examples.
```
Comment (jfrery) [inline docs/built-in-models/training.md:46]:
```suggestion
The `SGDClassifier` does not currently support training models with other values for the bit-widths. Second, the execution time of a single iteration
is proportional to the number of features and the number of training samples in the batch.
```

**Review by jfrery — CHANGES_REQUESTED** [review]
Thanks a lot! Just a few comments

**Review by bcm-at-zama — COMMENTED** [review]
Thanks for the different changes, in the doc and in the nb

What about adding also a short sentence / entry in https://github.com/zama-ai/concrete-ml#online-demos-and-tutorials for FL example and for FHE training? Really to me, the #online-demos-and-tutorials is an excellent entry point for our demos / showskills, so I would try to have it up to date with our best capabilities.

Possibly, if you feel there are too many things in #online-demos-and-tutorials, you could reduce it a bit, but I would keep the best things there

File: docs/built-in-models/training.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+Next, to perform the training on encrypted data, call the `fit` function with the `fhe="execute"` argument:
+
+<!--pytest-codeblocks:skip-->
+
+```python
+sgd_clf_binary_fhe.fit(X_binary, y_binary, fhe="execute")
+```
+
+{% hint style="info" %}
+Training on encrypted data provides the highest level of privacy but is slower than training on clear data. Federated learning is an alternative approach, where data privacy can be ensured through _differential privacy_ instead of encryption. Concrete ML
```
[the remark this answers is not in the document]
Comment (andrei-stoian-zama) [reply docs/built-in-models/training.md:30]: I think we should talk about it. People care about PPML not specific solutions, if we can show how FHE and FL/DP are complementary then we should.

I mention DP because that's what brings the privacy to FL. Do you not agree?

File: docs/built-in-models/training.md (hunk 5 above)
Comment (andrei-stoian-zama) [reply docs/built-in-models/training.md:37]: error on my part

File: docs/built-in-models/training.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (jfrery) [reply docs/built-in-models/training.md:30]: > I mention DP because that's what brings the privacy to FL

FL is a privacy training tech as is. Not sure how DP is is related here.

File: docs/built-in-models/training.md (hunk 4 above)
Comment (bencrts) [reply docs/built-in-models/training.md:45]: I agree -- I think it would read fine with it removed:
`The execution time of a single iteration
is proportional to the number of features and the number of training samples in the batch.`

File: docs/built-in-models/training.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (andrei-stoian-zama) [reply docs/built-in-models/training.md:30]: ok, good point

File: README.md [hunk 8]
```diff
@@ -154,6 +154,8 @@ Various tutorials are given for [built-in models](docs/built-in-models/ml_exampl
 - [Health diagnosis](use_case_examples/disease_prediction/): based on a patient's symptoms, history and other health factors, give
   a diagnosis using FHE to preserve the privacy of the patient.
 
+- [Private inference for federated learned models](use_case_examples/federated_learning/): private training of a Logistic Regression model and then import the model into Concrete ML and perform encrypted prediction
```
Comment (andrei-stoian-zama) [inline README.md:157]: @bcm-at-zama added this as you suggested

File: docs/advanced-topics/advanced_features.md [hunk 9]
```diff
@@ -6,25 +6,25 @@ Concrete ML provides features for advanced users to adjust cryptographic paramet
...
+The result of TLU operations is obtained with a specific tolerance to off-by-one errors. Concrete ML offers the possibility to set the probability of such errors occurring, which influences the cryptographic parameters. The lower the tolerance, the more restrictive the parameters become, making both key generation and, more significantly, FHE execution time slower.
 
 {% hint style="info" %}
 Concrete ML has a _simulation_ mode where the impact of approximate computation of TLUs on the model accuracy can be determined. The simulation is much faster, speeding up model development significantly. The behavior in simulation mode is representative of the behavior of the model on encrypted data.
 {% endhint %}
 
-In Concrete ML, there are three different ways to define the error probability:
+In Concrete ML, there are three different ways to define the  tolerance to off-by-one errors for each TLU operation:
```
Comment (RomanBredehoft) [inline docs/advanced-topics/advanced_features.md:15]: double space in `the  tolerance` !

File: docs/advanced-topics/advanced_features.md [hunk 10]
```diff
@@ -6,25 +6,25 @@ Concrete ML provides features for advanced users to adjust cryptographic paramet
...
 {% endhint %}
 
-### An error probability for an individual TLU
+### Tolerance to off-by-one error for an individual TLU
 
-The first way to set error probabilities in Concrete ML is at the local level, by directly setting the probability of error of each individual TLU. This probability is referred to as `p_error`. A given PBS operation has a `1 - p_error` chance of being successful. The successful evaluation here means that the value decrypted after FHE evaluation is exactly the same as the one that would be computed in the clear.
+The first way to set error probabilities in Concrete ML is at the local level, by directly setting the tolerance to error of each individual TLU operation (such as activation functions for a neuron output). This tolerance is referred to as `p_error`. A given PBS operation has a `1 - p_error` chance of being correct 100% of the time. The successful evaluation here means that the value decrypted after FHE evaluation is exactly the same as the one that would be computed in the clear. Otherwise, off-by-one errors might occur, which may not actually reduce model accuracy in practice.
```
Comment (RomanBredehoft) [inline docs/advanced-topics/advanced_features.md:27]: > Otherwise, off-by-one errors might occur, which may not actually reduce model accuracy in practice

should we explain why is that ? feels a bit weird to mention it just like this. Since we're only talking about off-by-one error for an _individual_ TLU I would maybe remove the `which may not actually reduce model accuracy in practice.` part

File: docs/built-in-models/training.md [hunk 11]
```diff
@@ -0,0 +1,47 @@
+# Training on Encrypted Data
+
+Concrete ML offers the possibility to train [SGD Logistic Regression](../developer-guide/api/concrete.ml.sklearn.linear_model.md#class-sgdclassifier) on encrypted data. The [logistic regression training](../advanced_examples/LogisticRegressionTraining.ipynb) example shows this feature in action.
+
+This example shows how to instantiate a logistic regression model that trains on encrypted data:
+
+```python
+from concrete.ml.sklearn import SGDClassifier
+parameters_range = (-1.0, 1.0)
```
Comment (RomanBredehoft) [inline docs/built-in-models/training.md:9]: maybe we should also explain a bit what `parameters_range` refers to and why it's set to -1, 1 here

File: docs/built-in-models/training.md [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+{% hint style="info" %}
+Training on encrypted data provides the highest level of privacy but is slower than training on clear data. Federated learning is an alternative approach, where data privacy can be ensured by using a trusted gradient aggregator, coupled with optional _differential privacy_ instead of encryption. Concrete ML
+can import linear models, including logistic regression, that are trained using federated learning using the [`from_sklearn` function](linear.md#pre-trained-models).
+
+{% endhint %}
+
+## Training configuration
+
+The `max_iter` parameter controls the number of batches that are processed by the training algorithm.
```
Comment (RomanBredehoft) [inline docs/built-in-models/training.md:38]: this is a parameter from sklearn, I think we can remove it no ?

File: docs/deep-learning/fhe_assistant.md [hunk 13]
```diff
@@ -100,46 +105,37 @@ except RuntimeError as err:
...
 ```
 
-The error `table lookups are only supported on circuits with up to 16-bit integers` indicates that the 16-bit limit on the input of the Table Lookup operation has been exceeded. To pinpoint the model layer that causes the error, Concrete ML provides the [bitwidth_and_range_report](../developer-guide/api/concrete.ml.quantization.quantized_module.md#method-bitwidth_and_range_report) helper function. First, the model must be compiled so that it can be [simulated](#simulation). Then, calling the function on the module above returns the following:
+The error  `this 17-bit value is used as an input to a table lookup` indicates that the 16-bit limit on the input of the Table Lookup (TLU) operation has been exceeded. To pinpoint the model layer that causes the error, Concrete ML provides the [bitwidth_and_range_report](../developer-guide/api/concrete.ml.quantization.quantized_module.md#method-bitwidth_and_range_report) helper function. First, the model must be compiled so that it can be [simulated](#simulation).
+
+To make this network FHE-compatible one can apply several techniques:
```
Comment (RomanBredehoft) [inline docs/deep-learning/fhe_assistant.md:121]: I would add a title around here to emphasize on the solutions

File: docs/deep-learning/fhe_assistant.md [hunk 14]
```diff
@@ -100,46 +105,37 @@ except RuntimeError as err:
...
-    '/fc1/Gemm': {'range': (-6180, 6840), 'bitwidth': 14}, 
-    '/fc2/Gemm': {'range': (-45051, 43090), 'bitwidth': 17}, 
-    '/fc3/Gemm': {'range': (-17351, 13868), 'bitwidth': 16}
-}
-```
-
-To make this network FHE-compatible one can reduce the bit-width of the second layer named `fc2`. To do this, a simple solution is to reduce the number of neurons, as it is proportional to the bit-width.
 
-Reducing the number of neurons in this layer resolves the error and makes the network FHE-compatible:
+2. reduce the accumulator bit-width of the second layer named `fc2`. To do this, a simple solution is to reduce the number of neurons, as it is proportional to the bit-width.
```
Comment (RomanBredehoft) [inline docs/deep-learning/fhe_assistant.md:138]: should we also talk about playing around the p_error ?

File: docs/advanced-topics/advanced_features.md (hunk 10 above)
Comment (andrei-stoian-zama) [reply docs/advanced-topics/advanced_features.md:27]: I rephrased. I think it's important to say that these are not "errors" per se.

Comment (RomanBredehoft) [reply docs/advanced-topics/advanced_features.md]: yes great thanks !

File: docs/built-in-models/training.md (hunk 12 above)
Comment (RomanBredehoft) [reply docs/built-in-models/training.md:38]: I would maybe emphasize on that then ! saying that's it's important for whatever reasons

**Review by RomanBredehoft — APPROVED** [review]
many thanks ! I would maybe emphasize on the max_iter parameter but other than that fine with me

File: docs/built-in-models/training.md [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+
+The `parameters_range` parameter determines the initialization of the coefficients and the bias of the logistic regression. It is recommended to give values that are close to the min/max of the training data. It is also possible to normalize the training data so that it lies in the range $$[-1, 1]$$.
```
Comment (fd0r) [inline docs/built-in-models/training.md:40]: Maybe we should add that this isn't just for initialization but also for quantization?

**Review by fd0r — APPROVED** [review]
lgtm

File: docs/advanced-topics/advanced_features.md [hunk 16]
```diff
@@ -6,25 +6,25 @@ Concrete ML provides features for advanced users to adjust cryptographic paramet
...
+The result of TLU operations is obtained with a specific tolerance to off-by-one errors. Concrete ML offers the possibility to set the probability of such errors occurring, which influences the cryptographic parameters. The lower the tolerance, the more restrictive the parameters become, making both key generation and, more significantly, FHE execution time slower.
 
 {% hint style="info" %}
 Concrete ML has a _simulation_ mode where the impact of approximate computation of TLUs on the model accuracy can be determined. The simulation is much faster, speeding up model development significantly. The behavior in simulation mode is representative of the behavior of the model on encrypted data.
 {% endhint %}
 
-In Concrete ML, there are three different ways to define the error probability:
+In Concrete ML, there are three different ways to define the tolerance to off-by-one errors for each TLU operation:
```
[the remark this answers is not in the document]
Comment (andrei-stoian-zama) [reply docs/advanced-topics/advanced_features.md:15]: no.. way to complicated to explain

**Review by RomanBredehoft — APPROVED** [review]

File: docs/README.md [hunk 17]
```diff
@@ -4,10 +4,13 @@
 
 ![](.gitbook/assets/3.png)
 
-Concrete ML is an open source, privacy-preserving, machine learning inference framework based on Fully Homomorphic Encryption (FHE). It enables data scientists without any prior knowledge of cryptography to automatically turn machine learning models into their FHE equivalent, using familiar APIs from scikit-learn and PyTorch (see how it looks for [linear models](built-in-models/linear.md), [tree-based models](built-in-models/tree.md), and [neural networks](built-in-models/neural-networks.md)).
+Concrete ML is an open source, privacy-preserving, machine learning framework based on Fully Homomorphic Encryption (FHE). It enables data scientists without any prior knowledge of cryptography to automatically turn machine learning models into their FHE equivalent, using familiar APIs from scikit-learn and PyTorch (see how it looks for [linear models](built-in-models/linear.md), [tree-based models](built-in-models/tree.md), and [neural networks](built-in-models/neural-networks.md)). Concrete ML supports converting models for inference with FHE but can also [train some models](built-in-models/training.md) on encrypted data.
```
Comment (kcelia) [inline docs/README.md:7]: The sentence `Concrete ML supports converting models for inference with FHE but can also [train some models](built-in-models/training.md) on encrypted data.` is  grammatically correct, however, the use of "but" might be slightly misleading, as the two features (model conversion for inference and model training on encrypted data) are not in opposition but are complementary

Comment (kcelia) [reply docs/README.md:7]: I would suggest: 
Concrete ML not only supports converting models for inference with FHE, it also enables training somes models ...
